iModelCore/ecobjects/src/ECProperty.cpp

```diff
@@ -1397,10 +1397,15 @@ ECObjectsStatus PrimitiveECProperty::SetType (PrimitiveType primitiveType)
 +---------------+---------------+---------------+---------------+---------------+------*/
 ECObjectsStatus PrimitiveECProperty::SetType (ECEnumerationCR enumerationType)
     {
-    if (&(enumerationType.GetSchema()) != &(this->GetClass().GetSchema()))
+    ECSchemaCR enumSchema = enumerationType.GetSchema();
+    ECSchemaCR currentSchema = this->GetClass().GetSchema();
+    if (&enumSchema != &currentSchema)
         {
-        if (!ECSchema::IsSchemaReferenced(this->GetClass().GetSchema(), enumerationType.GetSchema(), ECN::SchemaMatchType::Exact, true))
+        if (!ECSchema::IsSchemaReferenced(currentSchema, enumSchema) && ECObjectsStatus::Success != const_cast<ECSchemaR>(currentSchema).AddReferencedSchema(const_cast<ECSchemaR>(enumSchema)))

```

Comment:
The Set method should not add the schema reference, it should only check and return an error if the reference is not found.

This is the consistent behavior for all SetType overloads (Enumeration, struct and struct array), add base class, adding a custom attribute (except for deserialization to handle schemas created w/o property references to shipped bugs), setting a properties category, KOQ, nav prop relationship, etc.

It's a bit of a pain but it is worse if we are inconsistent.  So the calling code will have to check and add the reference if it is necessary.

Also class also has a GetSchemaR method to get the non const reference ... it just does the const_cast but it looks cleaner on the calling side.

Reply:
I have updated the code based on your feedback.